File: mortgage-iq-app/src/orchestrator/app/connectors/web_iq.py
"""Web IQ - web-scale grounding (Grounding with Bing).

Currently returns a canned market snapshot; the value is lightly varied by loan
size so different applications show different rate context. A future phase wires
this to a real Grounding-with-Bing connection on the Foundry agent.
"""

from __future__ import annotations

from typing import Any


class WebIQ:
    name = "web"
    display = "Web IQ"
    source = "Grounding with Bing"

    async def run(self, query: str, loan_amount: int = 0) -> dict[str, Any]:
        # Small deterministic variation so the demo does not look static.
        base = 6.52
        adjust = 0.05 if loan_amount and loan_amount > 500000 else 0.0
        return {
            "detail": "Checking today's market rates and TRID disclosure rules",
            "live": False,
            "data": {
                "avg_30yr_fixed": round(base + adjust, 2),
                "avg_15yr_fixed": round(base + adjust - 0.78, 2),
                "rate_trend": "down 0.08% week-over-week",
                "regulatory_note": "TRID: Loan Estimate must be delivered within 3 business days.",
            },
        }

File: mortgage-iq/lib/iq/web-iq.ts
/** Web IQ - current mortgage market rates + regulatory context (Grounding with Bing / public API). */

import { WEB_IQ_RATES_URL } from "../constants";
import type { IqToolResult, WebIqSnapshot } from "../types";

const SNAPSHOT: WebIqSnapshot = {
  avg_30yr_fixed: 6.52,
  avg_15yr_fixed: 5.74,
  rate_trend: "down 0.08% week-over-week",
  regulatory_note: "TRID: Loan Estimate must be delivered within 3 business days.",
};

export async function getWebIq(
  _query: string,
  loanAmount = 0,
): Promise<IqToolResult<WebIqSnapshot>> {
  if (WEB_IQ_RATES_URL) {
    try {
      const res = await fetch(WEB_IQ_RATES_URL, { cache: "no-store" });
      if (res.ok) {
        const body = (await res.json()) as Partial<WebIqSnapshot>;
        return {
          data: { ...SNAPSHOT, ...body },
          detail: "Fetched current market rates from the web",
          live: true,
        };
      }
    } catch (err) {
      console.warn("[web-iq] rates fetch failed:", err instanceof Error ? err.message : err);
    }
  }
  // Small deterministic variation so the snapshot is not perfectly static.
  const adjust = loanAmount > 500000 ? 0.05 : 0;
  return {
    data: {
      ...SNAPSHOT,
      avg_30yr_fixed: Math.round((SNAPSHOT.avg_30yr_fixed + adjust) * 100) / 100,
      avg_15yr_fixed: Math.round((SNAPSHOT.avg_15yr_fixed + adjust) * 100) / 100,
    },
    detail: "Checking today's market rates and TRID disclosure rules",
    live: false,
  };
}
